client/py/synnax/io/tdms.py

```diff
@@ -0,0 +1,144 @@
+
+from nptdms import TdmsFile, TdmsGroup, TdmsChannel
+
+from math import ceil
+
+from synnax.io.matcher import new_extension_matcher
+from synnax.io.protocol import ChannelMeta, ReaderType, File, BaseReader, ColumnFileReader
+
+TDMSMatcher = new_extension_matcher(["tdms"])
+
+
+class TDMSReader(TDMSMatcher):   # type: ignore
+    """A ColReader implementation for TDMS files."""
+    
+    channel_keys: set[str]
+    chunk_size: int
+
+    _current_chunk: int
+    _n_chunks: int | None
+    
+    _path: Path
+    _channels: list[ChannelMeta] | None
+
+    # Doing a protocol implementation check here because
+    # it's hard for pyright to handle factories that return
+    # protocol classes.
+    def _(self) -> ColumnFileReader:
+        return self
+
+    def __init__(
+        self,
+        path: Path,
+        keys: list[str] | None = None,
+        chunk_size: int | None = None,
+    ):
+        self._path = path
+        self.chunk_size = chunk_size or int(5e5)
+        
+        self._channels = None
+        self._current_chunk = 0
+        self._n_chunks = None
+        if keys is not None:
+            self.channel_keys = set(keys)
+        else:
+            self.set_keys_from_file()
+    
+    def channels(self) -> list[ChannelMeta]:
+        """:returns : a list of channel metadata for the file."""
+        if self._channels is not None:
+            return self._channels
+        
+        self._channels: list[ChannelMeta] = list()
+        with TdmsFile.open(self._path) as tdms_file:
+            for group in tdms_file.groups():
+                for channel in group.channels():
+                    self._channels.append(ChannelMeta(name=channel.name, meta_data=dict()))
+                    
+        return self._channels
+    
+    def set_chunk_size(self, chunk_size: int):
+        self.chunk_size = chunk_size
+    
+    def set_keys(self, keys: list[str]):
+        self.channel_keys = set(keys)
+    
+    def set_keys_from_file(self) -> set[str]:
+        self.channel_keys = set([ch.name for ch in self.channels()])
+        return self.channel_keys
+
+    @classmethod
+    def type(cls) -> ReaderType:
+        """:returns : the type of reader."""
+        return ReaderType.Column
+
+    def path(self) -> Path:
+        """:returns: the path to the file."""
+        return self._path
+
+    def nsamples(self) -> int:
+        """:returns: the number of samples in the file."""
+        return self.chunk_size * self.n_chunks * len(self.channels())
+
+    def seek_first(self):
+        """Seeks the reader to the first sample in the file."""
+        self._current_chunk = 0
+    
+    def read(self, *keys: str) -> pd.DataFrame:
+        """Reads a dataframe with chunk size * number of columns samples. The returned
+        dataframe contains columns for each key in keys.
+        """ 
+        # if we already read everything, return an empty chunk
+        if self._current_chunk >= self.n_chunks:
+            return pd.DataFrame()
```

Comment:
We need some way to indicate when the read method is exhausted. In the csv reader implementation I raise a `StopIteration` error. This is a bit clunky as `read` might be used outside of an iterator. It could be reasonable to change the protocol to add a boolean indicator that says whether there is more data or not. Thoughts?